unblob/extractor.py

```diff
@@ -46,6 +47,45 @@ def fix_permissions(outdir: Path):
             path.chmod(0o664)
 
 
+def is_safe_path(basedir: Path, path: Path) -> bool:
+    try:
+        basedir.joinpath(path).resolve().relative_to(basedir.resolve())
+    except ValueError:
+        return False
+    return True
+
+
+def fix_symlink(path: Path, outdir: Path) -> Path:
+    """Fix symlinks by rewriting absolute symlinks to make them point within
+    the extraction directory (outdir), if it's not a relative symlink it is
+    either removed it it attempts to traverse outside of the extraction directory
+    or rewritten to be fully portable (no mention of the extraction directory
+    in the link value)."""
+    target = Path(os.readlink(path))
+
+    if target.is_absolute():
+        target = Path(target.as_posix().lstrip("/"))
+    else:
+        target = path.resolve()
+
+    safe = is_safe_path(outdir, target)
+
+    if not safe:
+        print(f"Path traversal attempt through symlink, removing {target}.")
```

Comment:
I agree with you that we should create a Report about that issue (e.g. `MaliciousSymlinkRemoved`) and log it, but I think we should still get rid of it. It's a really weird edge case where someone crafted a file with links like '../../../../test/file` given that all the other links were normalized (even `/etc/passwd`), so we must delete that file.

However, this brings up another subject we need to explore in more depth with the team: 7z handling of 'malicious' symlinks.

7z delete any symlink that points outside of the extraction directory and we don't have control over that behavior (no command line flag to force it to create those links). It does that for any format it can extract. It's great and we follow its behavior by deleting such symlinks that may have been kept by other extractors.

However, 7z validation somehow acts differently.

It will delete links pointing to files that do not exists on disk:

```
ERROR: Dangerous link path was ignored : etc/resolv.conf : /mnt/tmp/resolv.conf
ERROR: Dangerous link path was ignored : lib/modules/4.9.44/build : /home/mdq/new_svn/sys/nt98312/NA51039_BSP_V1.01/BSP/linux-kernel
```

It will delete links having a '../' traversal in them, even if it's entirely valid. To me the symlink below should have been created:

```
ERROR: Dangerous symbolic link path was ignored : usr/bin/awk : ../../bin/busybox
```

I'll investigate how 7z acts in more details and document it. We should then take a decision: either we keep 7z as is and accept how it works, or fork and remove the check (I don't like that idea given the burden of keeping the repo up to date and potentially performing some backporting).

Fun times ahead !

Reply:
I'll go through all my logs next week to identify which formats push 7z to simply unlink "dangerous" link paths. I also want to check for each format whether we should move to something else like patool (https://github.com/wummel/patool), which seems to not really care about those.

Will keep you posted !

Reply:
I just wanted to chime saying that it is extremely easy using Nix to apply out-of-tree patches to any dependency without much hassle.